lib/catalog/owid/catalog/utils.py

```diff
@@ -268,3 +269,55 @@ def hash_any(x: Any) -> int:
     else:
         # Fallback for other types: use the built-in hash() function
         return hash(x)
+
+
+def dataclass_from_dict(cls: Optional[Type[T]], d: Dict[str, Any]) -> T:
+    """Recursively create an instance of a dataclass from a dictionary."""
+    if d is None or not dataclasses.is_dataclass(cls) or not isinstance(d, dict):
+        return d  # type: ignore
+
+    field_types = {f.name: f.type for f in dataclasses.fields(cls)}
+
+    init_args = {}
+    for field_name, v in d.items():
+        # Skip values in a dictionary that are not in the dataclass
+        if field_name not in field_types:
+            continue
+
+        # Handle None values right away
+        if v is None:
+            init_args[field_name] = None
+            continue
+
+        field_type = field_types[field_name]
+        origin = get_origin(field_type)
+        args = get_args(field_type)
+
+        # unwrap Optional (e.g. Optional[License] -> License)
+        if type(None) in args:
+            filtered_args = tuple(a for a in args if a is not type(None))
+            if len(filtered_args) == 1:
+                field_type = filtered_args[0]
+
+        if origin is list:
+            item_type = args[0]
+            init_args[field_name] = [dataclass_from_dict(item_type, item) for item in v]
+        elif origin is dict:
+            key_type, value_type = args
+            init_args[field_name] = {k: dataclass_from_dict(value_type, item) for k, item in v.items()}
+        elif dataclasses.is_dataclass(field_type):
+            init_args[field_name] = dataclass_from_dict(field_type, v)  # type: ignore
+        elif isinstance(field_type, type):
+            try:
+                init_args[field_name] = field_type(v)
+            except ValueError as e:
+                # HACK 2024-11-08: we got invalid type into one of the fields, remove it when we fix it
+                # on all staging servers
+                if "invalid literal for int() with base 10: '2020" in str(e):
+                    init_args[field_name] = int(v.replace('"', ""))
```

Comment:
I'm not sure what's going on here. Maybe we should create an issue to remember to fix this properly?

Reply:
It comes [from a typo](https://github.com/owid/etl/pull/3517/files#diff-00e51537a22bb68140652296351986cdc3e1d430d97a27b103c4b83b62eb4280) in one old metadata file. I'll remove it the next time we increase ETL_EPOCH.